### PR TITLE
PR from fork/331

### PR DESCRIPTION
**Pull request by MichaelMcKinsey1**
PR from #331 Integrate Caliper

### CODE REVIEW

**Comment by rhornung67** [comment]
@MichaelMcKinsey1 I merged develop into this and resolved the merge conflicts. When checks pass, is this ready to merge?

**Comment by MichaelMcKinsey1** [comment]
> @MichaelMcKinsey1 I merged develop into this and resolved the merge conflicts. When checks pass, is this ready to merge?

1 more issue came up today. Once it is addressed we should be good to go.

**Comment by MichaelMcKinsey1** [comment]
> > @MichaelMcKinsey1 I merged develop into this and resolved the merge conflicts. When checks pass, is this ready to merge?
> 
> 1 more issue came up today. Once it is addressed we should be good to go.

Done in [51144fb](https://github.com/LLNL/RAJAPerf/pull/352/commits/51144fb282f31aac4f02f0970f6eb3470dc2d07e).  

Also, @daboehme wanted to cleanup the caliper implementation before the merge, so that will come soon.

**Comment by rhornung67** [comment]
Sounds good. I review and get some others to also review when you're ready. Please let me know.

**Comment by MichaelMcKinsey1** [comment]
> > > @MichaelMcKinsey1 I merged develop into this and resolved the merge conflicts. When checks pass, is this ready to merge?
> > 
> > 
> > 1 more issue came up today. Once it is addressed we should be good to go.
> 
> Done in [51144fb](https://github.com/LLNL/RAJAPerf/pull/352/commits/51144fb282f31aac4f02f0970f6eb3470dc2d07e).
> 
> Also, @daboehme wanted to cleanup the caliper implementation before the merge, so that will come soon.

#359

**Comment by MichaelMcKinsey1** [comment]
@rhornung67 Changes merged. Ready now.

File: src/algorithm/MEMCPY-Cuda.cpp [hunk 1]
```diff
@@ -157,7 +157,7 @@ void MEMCPY::runCudaVariant(VariantID vid, size_t tune_idx)
         run_params.validGPUBlockSize(block_size)) {
 
       if (tune_idx == t) {
-
+        setBlockSize(block_size);
```
Comment (rhornung67) [inline src/algorithm/MEMCPY-Cuda.cpp:160]: @MrBurmark please take a look at these changes and indicate whether you have any questions or concerns.

**Review by rhornung67 — COMMENTED** [review]
A couple of minor comments.

File: scripts/lc-builds/blueos_nvcc_clang_caliper.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  -C ${RAJA_HOSTCONFIG} \
+  -DENABLE_OPENMP=On \
+  -DENABLE_CUDA=On \
+  -DCUDA_TOOLKIT_ROOT_DIR=/usr/tce/packages/cuda/cuda-${COMP_NVCC_VER} \
+  -DCMAKE_CUDA_COMPILER=/usr/tce/packages/cuda/cuda-${COMP_NVCC_VER}/bin/nvcc \
+  -DCMAKE_CUDA_ARCHITECTURES=${COMP_ARCH} \
+  -DCMAKE_INSTALL_PREFIX=../install_${BUILD_SUFFIX} \
+  -DRAJA_PERFSUITE_USE_CALIPER=ON \
+  -Dcaliper_DIR=/usr/workspace/wsb/asde/caliper-lassen/share/cmake/caliper \
+  -Dadiak_DIR=/usr/workspace/wsb/asde/adiak-lassen/lib/cmake/adiak \
```
Comment (MrBurmark) [inline scripts/lc-builds/blueos_nvcc_clang_caliper.sh:54]: Should the script take a path to these libraries to make them more general?

File: src/algorithm/MEMCPY-Cuda.cpp (hunk 1 above)
Comment (MrBurmark) [reply src/algorithm/MEMCPY-Cuda.cpp:160]: This particular change is to allow capture of the block size as metadata. I'm not thrilled with this but I'm not sure there is a better way to do it either. I have some ideas about refactoring tunings that may improve things a bit but that's longer term.

File: src/common/Executor.cpp [hunk 3]
```diff
@@ -108,6 +119,99 @@ Executor::Executor(int argc, char** argv)
...
+
+  // Openmp section
+#if defined(_OPENMP)
+  std::string strval = "";
+  std::string test = std::to_string(_OPENMP);
+
+  std::unordered_map<unsigned,std::string> map{
+    {200505,"2.5"},{200805,"3.0"},{201107,"3.1"},{201307,"4.0"},{201511,"4.5"},{201611,"4.5"},{201811,"5.0"},{202011,"5.1"},{202111,"5.2"}};
+
+  try {
```
Comment (MrBurmark) [inline src/common/Executor.cpp:203]: Use find instead of try catch.

File: src/common/Executor.cpp [hunk 4]
```diff
@@ -670,7 +777,15 @@ void Executor::setupSuite()
 
           tuning_names[vid].resize(tuning_names_order_map.size());
           for (auto const& tuning_name_idx_pair : tuning_names_order_map) {
-            tuning_names[vid][tuning_name_idx_pair.second] = tuning_name_idx_pair.first;
+            size_t tid = tuning_name_idx_pair.second;
+            std::string tstr = tuning_name_idx_pair.first;
```
Comment (MrBurmark) [inline src/common/Executor.cpp:781]:
```suggestion
            std::string const& tstr = tuning_name_idx_pair.first;
```

File: src/algorithm/MEMCPY-Cuda.cpp (hunk 1 above)
Comment (MichaelMcKinsey1) [reply src/algorithm/MEMCPY-Cuda.cpp:160]: Most kernels' block size are being set by a change in `src/common/GPUUtils.hpp::RAJAPERF_GPU_BLOCK_SIZE_TUNING_DEFINE_BOILERPLATE`. But some like `MEMCPY-Cuda.cpp` don't contain the boilerplate so I had to put them in manually like above. To clarify this was to get per kernel block size and not for the run metadata.

File: scripts/lc-builds/blueos_nvcc_clang_caliper.sh (hunk 2 above)
Comment (MichaelMcKinsey1) [reply scripts/lc-builds/blueos_nvcc_clang_caliper.sh]: [011643d](https://github.com/LLNL/RAJAPerf/pull/352/commits/011643dce385978d67131478e8b664bbb4d0c5bd)

File: src/common/Executor.cpp (hunk 3 above)
Comment (MichaelMcKinsey1) [reply src/common/Executor.cpp]: [962334b](https://github.com/LLNL/RAJAPerf/pull/352/commits/962334b53464aa150e3a9ac9c44343071430943c)

**Review by rhornung67 — APPROVED** [review]

**Review by MrBurmark — APPROVED** [review]

**Comment by rhornung67** [comment]
@MichaelMcKinsey1 I merged develop into this. There were several merge conflicts in Executor.cpp that I resolved because I refactored Executor::setupSuite(). When you have time, please take a look to make sure I didn't screw up anything you did. When you are OK with it, I will merge it. Thank you.

**Comment by rhornung67** [comment]
No worries.

**Comment by MichaelMcKinsey1** [comment]
@rhornung67 Fixed and re-merged develop.

**Comment by rhornung67** [comment]
> @rhornung67 Fixed and re-merged develop.

Thanks for looking over @MichaelMcKinsey1 I will merge this when checks come back green. It may be a couple of days since our ruby platform is down sue to cooling issues.

**Pull request merged**